src/bank02.asm

```diff
@@ -8130,29 +8126,194 @@ titleScreenLicenseText:
-    TXT  "    the world.<00>"                          ;; 02:7fd8 ???????????????
-    TXT  "<00>"                                        ;; 02:7fe7 ?
-    db   $00, $00, $00, $00, $00, $00, $00, $00        ;; 02:7fe8 ????????
-    db   $00, $00, $00, $00, $01, $ff, $ff, $ff        ;; 02:7ff0 ????????
-    db   $ff, $ff, $ff, $ff, $ff, $ff, $ff, $ff        ;; 02:7ff8 ????????
+    TXT  "<00>"                                        ;; 02:7fd8 ?
+    db   $01                                           ;; 02:7fd9 ?
+
+; Draws left aligned number HL as modified tiles at WRAM offset E
+; D holds the drawing mode in the first nibble (0 for shift, 1 for shift/swap)
+; D holds number of max digits in the second nibble
+drawLeftAlignedNumberInWRAM:
+    ld   B, $00 ; digit counter
+    ld   C, D ; use C to hold mode/max-digit-count
+
+    ; Housekeeping, get a copy of proper WRAM starting
+    ;  location on the stack for later use
+    push HL
+    ld   HL, wHPMPTileBuffer
+    ld   D, B
+    add  HL, DE
+    ld   D, H
+    ld   E, L
+    pop  HL
+    push BC
+    push DE
+
+    ; Load up the digits on the stack
+.divmod_and_store_number:
+    ld   A, $0a
+    push BC
+    push DE
+    call divMod
+    pop  DE
+    pop  BC
+    push AF
+    inc  B
+    ld   A, H
+    or   A, L
+    jr   NZ, .divmod_and_store_number
+
+    ; Reduce second nibble of C by num digits found
+    ld   A, C
+    sub  A, B
+    ld   C, A
+
+    ; Increment DE if in swap mode
+    ld   H, D
+    ld   L, E
+    ld   D, $00
+    and  A, $f0
+    ld   E, A
+    add  HL, DE
+    ld   D, H
+    ld   E, L
+
+    ; Stack holds B digits to pop, transfer them
+ .grab_digit_and_transfer:
+    ld   HL, $6800
+    pop  AF ; grab digit
+    swap A
+    push DE
+    ld   D, $00
+    ld   E, A
+    add  HL, DE
+    pop  DE
+    ld   A, $08
+    push BC
+    ld   B, $10
+    call copyShiftBankAfromHLtoDE
+    pop  BC
+    dec  B
+    jr   NZ, .grab_digit_and_transfer
+
+    ; Clear remaining tiles
+    ld   A, C
+    and  A, $0f
+    jr   Z, .check_swap
+
+    ld   H, D
+    ld   L, E
+    ld   B, A
+    xor  A, A
+ .clear_remaining_tiles:
+    call memsetTileWithA
+    dec  B
+    jr   NZ, .clear_remaining_tiles
+
+    ; Shifted tiles are in WRAM, check if swap needed
+ .check_swap:
+    pop  DE
+    ld   A, C
+    ld   L, C
+    pop  BC
+    and  A, $f0
+    ret  Z
+
+    ; We need to swap back tiles, prepare...
+    ld   A, C
+    sub  A, L
+    ld   B, A ; B holds tiles to swap
+
+    jp   swapBackHalfTiles
+
+; Swaps tiles back across half tile boundaries
+; B holds number of tiles to shift
+; DE holds address of first tile
+swapBackHalfTiles:
+    swap B ; now B holds bytes to swap
+
+    ; Clear the first tile (it was left untouched so far)
+    ld   H, D
+    ld   L, E
+    xor  A, A
+    call memsetTileWithA
+
+.shift_4px:
+    ld   A, [HL]
+    swap A
+    and  A, $0f
+    ld   C, A
+    ld   A, [DE]
+    or   A, C
+    ld   [DE], A
```

Comment:
One question I have is, is there an advantage to swapping what's in `hl` and what is in `de` (in this function and/or in `drawLeftAlignedNumberInWRAM`)? I'm not sure yet, but this could be:
`or [hl]`
`ld [hl+], c`

Reply:
Yes! That takes the inner loop from 22 cycles to 18 cycles (this also allowed me to reuse C to store the swapped value in A).